### PR TITLE
feat: add repaint

### PR DESCRIPTION
**Pull request by Revist**
Added repaint pipeline based on https://arxiv.org/pdf/2201.09865.pdf @anton-l 

In tests i was not sure where you take numbers to compare against from so I run the pipeline and pasted what it produces for me: 
`[0.14537135, 0.10728511, 0.08822048, 0.15828621, 0.11806837, 0.11007798, 0.15231332, 0.1214554, 0.15475643]`

### CODE REVIEW

File: tests/pipelines/repaint/test_repaint.py
Comment (anton-l) [inline tests/pipelines/repaint/test_repaint.py:65]: Uploaded the test images and masks to streamline the test a bit. This should work once the PIL suggestion is applied.
Feel free to use different URLs if you make any changes, I'll rehost the images before the final merge :)

```suggestion
from diffusers import RePaintPipeline, RePaintScheduler, UNet2DModel
from diffusers.utils.testing_utils import load_image, require_torch_gpu, slow, torch_device


torch.backends.cuda.matmul.allow_tf32 = False


@slow
@require_torch_gpu
class RepaintPipelineIntegrationTests(unittest.TestCase):
    def test_celebahq(self):
        original_image = load_image(
            "https://huggingface.co/datasets/hf-internal-testing/diffusers-images/resolve/main/"
            "repaint/celeba_hq_256.png"
        )
        mask_image = load_image(
            "https://huggingface.co/datasets/hf-internal-testing/diffusers-images/resolve/main/repaint/mask_256.png"
        )
        expected_image = load_image(
            "https://huggingface.co/datasets/hf-internal-testing/diffusers-images/resolve/main/"
            "repaint/celeba_hq_256_result.png"
        )
        expected_image = np.array(expected_image, dtype=np.float32) / 255.0

        model_id = "google/ddpm-ema-celebahq-256"
        unet = UNet2DModel.from_pretrained(model_id)
        scheduler = RePaintScheduler.from_config(model_id)

        repaint = RePaintPipeline(unet=unet, scheduler=scheduler).to(torch_device)

        generator = torch.Generator(device=torch_device).manual_seed(0)
        output = repaint(
            original_image,
            mask_image,
            num_inference_steps=250,
            eta=0.0,
            jump_length=10,
            jump_n_sample=10,
            generator=generator,
            output_type="np",
        )
        image = output.images[0]

        assert image.shape == (256, 256, 3)
        assert np.abs(expected_image - image).mean() < 1e-2
```

File: src/diffusers/pipelines/repaint/pipeline_repaint.py
Comment (anton-l) [inline src/diffusers/pipelines/repaint/pipeline_repaint.py:85]: Added a bit of compatibility with PIL images and masks here, similar to the Stable Diffusion inpainting pipelines.

```suggestion
import numpy as np
import torch

import PIL
from tqdm.auto import tqdm

from ...models import UNet2DModel
from ...pipeline_utils import DiffusionPipeline, ImagePipelineOutput
from ...schedulers import RePaintScheduler


def _preprocess_image(image: PIL.Image.Image):
    image = np.array(image.convert("RGB"))
    image = image[None].transpose(0, 3, 1, 2)
    image = torch.from_numpy(image).to(dtype=torch.float32) / 127.5 - 1.0
    return image


def _preprocess_mask(mask: PIL.Image.Image):
    mask = np.array(mask.convert("L"))
    mask = mask.astype(np.float32) / 255.0
    mask = mask[None, None]
    mask[mask < 0.5] = 0
    mask[mask >= 0.5] = 1
    mask = torch.from_numpy(mask)
    return mask


class RePaintPipeline(DiffusionPipeline):
    unet: UNet2DModel
    scheduler: RePaintScheduler

    def __init__(self, unet, scheduler):
        super().__init__()
        self.register_modules(unet=unet, scheduler=scheduler)

    @torch.no_grad()
    def __call__(
        self,
        original_image: Union[torch.FloatTensor, PIL.Image.Image],
        mask: Union[torch.FloatTensor, PIL.Image.Image],
        num_inference_steps: int = 250,
        eta: float = 0.0,
        jump_length: int = 10,
        jump_n_sample: int = 10,
        generator: Optional[torch.Generator] = None,
        output_type: Optional[str] = "pil",
        return_dict: bool = True,
        **kwargs,
    ) -> Union[ImagePipelineOutput, Tuple]:
        r"""
        Args:
            original_image (`torch.FloatTensor` or `PIL.Image.Image`):
                The original image to inpaint on.
            mask (`torch.FloatTensor` or `PIL.Image.Image`):
                The mask where 0.0 values define which part of the original image to inpaint (change).
            num_inference_steps (`int`, *optional*, defaults to 1000):
                The number of denoising steps. More denoising steps usually lead to a higher quality image at the
                expense of slower inference.
            eta (`float`):
                The weight of noise for added noise in a diffusion step. Its value is between 0.0 and 1.0 - 0.0 is DDIM
                and 1.0 is DDPM scheduler respectively.
            jump_length (`int`, *optional*, defaults to 10):
                The number of steps taken forward in time before going backward in time for a single jump ("j" in
                RePaint paper). Take a look at Figure 9 and 10 in https://arxiv.org/pdf/2201.09865.pdf.
            jump_n_sample (`int`, *optional*, defaults to 10):
                The number of times we will make forward time jump for a given chosen time sample. Take a look at
                Figure 9 and 10 in https://arxiv.org/pdf/2201.09865.pdf.
            generator (`torch.Generator`, *optional*):
                A [torch generator](https://pytorch.org/docs/stable/generated/torch.Generator.html) to make generation
                deterministic.
            output_type (`str`, *optional*, defaults to `"pil"`):
                The output format of the generate image. Choose between
                [PIL](https://pillow.readthedocs.io/en/stable/): `PIL.Image.Image` or `np.array`.
            return_dict (`bool`, *optional*, defaults to `True`):
                Whether or not to return a [`~pipeline_utils.ImagePipelineOutput`] instead of a plain tuple.

        Returns:
            [`~pipeline_utils.ImagePipelineOutput`] or `tuple`: [`~pipelines.utils.ImagePipelineOutput`] if
            `return_dict` is True, otherwise a `tuple. When returning a tuple, the first element is a list with the
            generated images.
        """

        if not isinstance(original_image, torch.FloatTensor):
            original_image = _preprocess_image(original_image)
        original_image = original_image.to(self.device)
        if not isinstance(mask, torch.FloatTensor):
            mask = _preprocess_mask(mask)
        mask = mask.to(self.device)
```

File: src/diffusers/pipelines/repaint/pipeline_repaint.py
Comment (anton-l) [inline src/diffusers/pipelines/repaint/pipeline_repaint.py:91]: Since the generators are device-specific all over the library now (instead of always generating the noise on CPU), this makes sure that we can use a cuda generator.

```suggestion
        image = torch.randn(
            original_image.shape,
            generator=generator,
            device=self.device,
        )
```

File: src/diffusers/schedulers/scheduling_repaint.py
Comment (anton-l) [inline src/diffusers/schedulers/scheduling_repaint.py:276]: Same as above, adding cuda support

```suggestion
        noise = torch.randn(
            model_output.shape, dtype=model_output.dtype, generator=generator, device=model_output.device
        )
```

File: src/diffusers/schedulers/scheduling_repaint.py
Comment (anton-l) [inline src/diffusers/schedulers/scheduling_repaint.py:309]: And here too
```suggestion
            noise = torch.randn(sample.shape, generator=generator, device=sample.device)
```

**Review by anton-l — COMMENTED** [review]
Hi @Revist! Excited to see a working RePaint port, thank you so much for working on it! 

Overall the implementation is super nice, just left a couple of suggestions to bring it closer to the Stable Diffusion inpainting API (which got added after my original repaint PR).

Now I wonder if it's possible to move the noising & masking logic from `RePaintScheduler` to `RePaintPipeline`, so that it can be used with other schedulers like PNDM or K-LMS? By using just `step()` and `add_noise()`, similar to the unsupervised inpainting pipeline for Stable Diffusion: https://github.com/huggingface/diffusers/blob/main/src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_inpaint_legacy.py#L46
What do you think, would that be possible to do here?

File: src/diffusers/pipelines/repaint/pipeline_repaint.py
Comment (anton-l) [reply src/diffusers/pipelines/repaint/pipeline_repaint.py:85]: Oof, sorry for such a big chunk of code, github doesn't even know how to render a diff for it :sweat_smile:

**Comment by Revist** [comment]
Hi @anton-l ! 

Was doing some reading as i was not that familiar with these, but it seems that although theoretically that should be possible (after all RePaint just repeats some steps) in practice you have some problems like here:
https://github.com/huggingface/diffusers/blob/main/src/diffusers/schedulers/scheduling_pndm.py#L209
where the step depends on some counter and not only on timestep itself. To include PNDM first thing would be to remove such dependency. 
Then one could build repaint timesteps and ask with step() and add_noise() to go forward and backward, where how to go forward would be recognized by timestep value and scheduler name.

**Comment by Revist** [comment]
Reading further in to make a step with PNDM you do not need a single model_output at single time but at t, t+delta/2 and t+delta so that changes logic here:
https://github.com/huggingface/diffusers/blob/main/src/diffusers/pipelines/pndm/pipeline_pndm.py#L85
where you would like to give a single model_output to step, you would have to pass model (unet) and image (x_t) instead.
You skipped this problem by creating virtual steps in this scheduler which are not really steps, then using this counter. But i would say the problems arise from this (and makes code complex imo).

**Comment by Revist** [comment]
@anton-l I am not sure if you agree with the above comments. If so I could take a look at how to change the code to the PNDM (and possibly other) schedulers however that seems like significant changes across code.

**Comment by anton-l** [comment]
@Revist I've tried refactoring it a bit and completely share your concerns now, let's leave the scheduler as is for now!

Pinging @patrickvonplaten @patil-suraj for a second review, as this is a new pipeline and scheduler :)

File: src/diffusers/pipelines/repaint/pipeline_repaint.py
Comment (patrickvonplaten) [inline src/diffusers/pipelines/repaint/pipeline_repaint.py:67]:
```suggestion
```

File: tests/pipelines/repaint/test_repaint.py
Comment (patrickvonplaten) [inline tests/pipelines/repaint/test_repaint.py:43]: Ok for now, but let's make sure to pass to numpy image tests in the very near future

**Review by patrickvonplaten — APPROVED** [review]
Looks good to me ! Before merging let's please make sure to add it to the docs as well and maybe reach out to the original author here?

File: tests/pipelines/repaint/test_repaint.py
Comment (anton-l) [reply tests/pipelines/repaint/test_repaint.py:43]: Will fetch the numpy images for all integration tests in a bulk PR :+1:

**Comment by anton-l** [comment]
Added the docs for the pipeline: https://moon-ci-docs.huggingface.co/docs/diffusers/pr_974/en/api/pipelines/repaint
And the scheduler: https://moon-ci-docs.huggingface.co/docs/diffusers/pr_974/en/api/schedulers#diffusers.RePaintScheduler

Gently pinging the RePaint author @andreas128 for an optional review and comments, in case out email thread gets lost :hugs:

**Review by patrickvonplaten — APPROVED** [review]

**Comment by anton-l** [comment]
@Revist the pipeline will be included in today's `diffusers==0.7.0` release! :partying_face: 

If you're interested, maybe we could work together on a Spaces demo similar to SD-inpainting https://huggingface.co/spaces/runwayml/stable-diffusion-inpainting to promote the pipeline further, let me know what you think! 
Also would be interesting to try integrating OpenAI's [improved-diffusion](https://github.com/openai/improved-diffusion) models to completely reproduce the RePaint paper :fire:

**Comment by Revist** [comment]
@anton-l Sounds interesting. What models would you like to integrate from improved diffusion? I was thinking of taking a look into https://github.com/huggingface/diffusers/issues/894 as this may be of use to our group in the near future. Maybe we could speak on some other communicator?

**Pull request merged**
